### PR TITLE
support live reloading of multiple pipelines

### PR DESCRIPTION
**Pull request by jsvd**
NOTE: this is a backport from master (pr #4520)

* reload config from files
* add reload config flag and allow reload through sighup
* move signal handling to runner
* add "-r" flag to detect changes in config and reload
* make SIGHUP reload config file
* make agent execute either loop to detect changes or sleep forever
* sigint/sigterm shut down the agent.execute task
* stop agent if pipeline terminates
* Better shutdown logic for controller and pipeline
* support multiple pipeline reloading and other cleanups
* serialize upgrade_state and running_pipelines?
* remove useless @thread from pipeline
* remove agent registry and cleanup
* reduce reload_interval from 5 to 3
* remove useless MissingAgentError class
* fix exit codes for reload_state failures
* explain why thread var exists in agent.rb
* properly pass backtrace exceptions to oops
* remove is_config_test from loader
* better documentation on agent/runner/pipeline methods

### CODE REVIEW

**Comment by jsvd** [comment]
@andrewvc review welcome on this port to 2.x branch

**Comment by suyograo** [comment]
@jsvd rebase?

**Comment by andrewvc** [comment]
@jsvd this now has merge conflicts (apologies for the late delay) can you rebase? One of them is with the patched 'oops' handling I added in master. We shouldn't be formatting the backtrace with I18N but rather passing it as a key to Lumberjack

File: logstash-core/lib/logstash/agent.rb [hunk 1]
```diff
@@ -3,12 +3,16 @@
 require "logstash/environment"
 require "logstash/errors"
 require "logstash/config/cpu_core_strategy"
+require "stud/trap"
+require "logstash/config/loader"
 require "uri"
 require "net/http"
 require "logstash/pipeline"
-LogStash::Environment.load_locale!
```
Comment (andrewvc) [inline logstash-core/lib/logstash/agent.rb]: Why was this removed?

Comment (jsvd) [inline logstash-core/lib/logstash/agent.rb]: I didn't see a need for it (didn't break anything). we still do `load_locale!` in 3 places:

logstash-core/lib/logstash/config/mixin.rb:LogStash::Environment.load_locale!
logstash-core/lib/logstash/environment.rb:    def load_locale!
logstash-core/lib/logstash/runner.rb:LogStash::Environment.load_locale!

Which should cover the paths to loading agent.rb

File: logstash-core/lib/logstash/agent.rb [hunk 2]
```diff
@@ -320,63 +296,195 @@ def configure_plugin_paths(paths)
...
+  end
+
+  def shutdown_pipelines
+    @pipelines.each { |id, _| stop_pipeline(id) }
+  end
+
+  def stop_pipeline(id)
+    pipeline = @pipelines[id]
+    return unless pipeline
+    @logger.warn("stopping pipeline", :id => id)
```
Comment (andrewvc) [inline logstash-core/lib/logstash/agent.rb]: Why is this logger.warn? Shouldn't this be at the info level?

File: logstash-core/lib/logstash/agent.rb [hunk 3]
```diff
@@ -320,63 +296,195 @@ def configure_plugin_paths(paths)
...
+  def reload_pipeline!(id)
+    old_pipeline = @pipelines[id]
+    new_pipeline = create_pipeline(old_pipeline.original_settings)
+    return if new_pipeline.nil?
+
+    if old_pipeline.config_str == new_pipeline.config_str
+      @logger.debug("no configuration change for pipeline",
+                    :pipeline => id, :config => old_pipeline.config_str)
+    else
+      @logger.warn("fetched new config for pipeline. upgrading..",
```
Comment (andrewvc) [inline logstash-core/lib/logstash/agent.rb]: Shouldn't this be logged at info, not warn?

File: logstash-core/lib/logstash/agent.rb (hunk 2 above)
Comment (jsvd) [inline logstash-core/lib/logstash/agent.rb]: True... By default we set the logger to only show >= warn and I thought this was an important enough message to show in normal usage.

Comment (andrewvc) [inline logstash-core/lib/logstash/agent.rb]: I think in that case we should do a logger.info + a logger.terminal. I don't think it should appear on-disk IMHO.

The real problem is that info should be safe to turn on by default IMHO. warn should be used for things that demand operator attention.

Comment (andrewvc) [inline logstash-core/lib/logstash/agent.rb]: What do you think @jordansissel ? How does this integrate with your long term plans for log level revision in LS?

Comment (jsvd) [inline logstash-core/lib/logstash/agent.rb]: an alternative for now is to just do a @logger.log, which will bypass the level checking, iirc.
I don't see a reason why this message shouldn't be in the disk logs

File: logstash-core/locales/en.yml [hunk 4]
```diff
@@ -5,7 +5,7 @@
 en:
   oops: |-
     The error reported is: 
-      %{error}
```
Comment (andrewvc) [inline logstash-core/locales/en.yml]: This will have a merge conflict now.

File: logstash-core/spec/logstash/agent_spec.rb [hunk 5]
```diff
@@ -1,39 +1,167 @@
...
 
-      context "unix" do
-        let(:path) { './test.conf' }
-        it 'works with relative path' do
-          subject.load_config(path)
+  before :each do
+    [:info, :warn, :error, :fatal, :debug].each do |level|
+      allow(logger).to receive(level)
+    end
+    [:info?, :warn?, :error?, :fatal?, :debug?].each do |level|
```
Comment (andrewvc) [inline logstash-core/spec/logstash/agent_spec.rb]: Maybe just have one array with both ? and non-? things?

File: logstash-core/lib/logstash/agent.rb (hunk 2 above)
Comment (andrewvc) [inline logstash-core/lib/logstash/agent.rb]: @jsvd I'm fine with `@logger.log`

File: logstash-core/lib/logstash/agent.rb (hunk 1 above)
Comment (andrewvc) [inline logstash-core/lib/logstash/agent.rb]: :+1:

**Comment by andrewvc** [comment]
Once the logger stuff is changed to @logger.log I'm LGTM here.

**Comment by jsvd** [comment]
closed by 0099537
